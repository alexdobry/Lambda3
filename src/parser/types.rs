use super::*;
use crate::lexer::SyntaxKind::*;

pub fn typ(p: &mut Parser) {
    let cp = p.checkpoint();
    if !atom(p) {
        p.report_error("expected type".to_string())
    }
    if p.at(Arrow) {
        p.finish_at(cp, TyArg);
        p.bump(Arrow);
        let cp2 = p.checkpoint();
        typ(p);
        p.finish_at(cp2, TyRes);
        p.finish_at(cp, FuncTy)
    }
}

fn atom(p: &mut Parser) -> bool {
    match p.current() {
        Int => {
            p.start_node(IntTy);
            p.bump(Int);
            p.finish_node();
        }
        Bool => {
            p.start_node(BoolTy);
            p.bump(Bool);
            p.finish_node()
        }
        LParen => {
            p.start_node(ParenTy);
            p.bump(LParen);
            typ(p);
            if !p.eat(RParen) {
                p.report_error("missing closing paren".to_string())
            }
            p.finish_node()
        }
        _ => return false,
    }
    true
}
