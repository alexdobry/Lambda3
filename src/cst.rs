use crate::lexer::SyntaxKind;
use crate::parser::{self, Parse};
use crate::syntax::{SyntaxNode, SyntaxToken};
use rowan::TextRange;

pub fn parse(text: &str) -> Parse {
    parser::parse(text)
}

pub fn parse_type(text: &str) -> Parse {
    parser::parse_type(text)
}

pub fn parse_pattern(text: &str) -> Parse {
    parser::parse_pattern(text)
}

macro_rules! ast_node {
    ($ast:ident, $kind:ident) => {
        #[derive(PartialEq, Eq, Hash, Debug)]
        #[repr(transparent)]
        pub struct $ast(pub SyntaxNode);
        impl $ast {
            #[allow(unused)]
            pub fn cast(node: SyntaxNode) -> Option<Self> {
                if node.kind() == SyntaxKind::$kind {
                    Some(Self(node))
                } else {
                    None
                }
            }
        }
    };
}

macro_rules! lit_node {
    ($ast:ident, $kind:ident) => {
        #[derive(PartialEq, Eq, Hash, Debug)]
        #[repr(transparent)]
        pub struct $ast(pub SyntaxNode);
        impl $ast {
            #[allow(unused)]
            fn cast(node: SyntaxNode) -> Option<Self> {
                if node.kind() == SyntaxKind::LiteralE
                    && first_token(&node)?.kind() == SyntaxKind::$kind
                {
                    Some(Self(node))
                } else {
                    None
                }
            }
        }
    };
    ($ast:ident, $kind1:ident, $kind2:ident) => {
        #[derive(PartialEq, Eq, Hash, Debug)]
        #[repr(transparent)]
        pub struct $ast(pub SyntaxNode);
        impl $ast {
            #[allow(unused)]
            fn cast(node: SyntaxNode) -> Option<Self> {
                if node.kind() == SyntaxKind::LiteralE {
                    let token = first_token(&node)?.kind();
                    if token == SyntaxKind::$kind1 || token == SyntaxKind::$kind2 {
                        Some(Self(node))
                    } else {
                        None
                    }
                } else {
                    None
                }
            }
        }
    };
}

ast_node!(Root, Root);
ast_node!(Var, VarE);
ast_node!(Lambda, LambdaE);
ast_node!(Application, ApplicationE);
lit_node!(IntLit, NumberLit);
lit_node!(BooleanLit, TrueKw, FalseKw);

#[derive(PartialEq, Eq, Hash, Debug)]
#[repr(transparent)]
pub struct Expr(SyntaxNode);

#[derive(Debug)]
pub enum ExprKind {
    Var(Var),
    Lambda(Lambda),
    Application(Application),
    IntLit(IntLit),
    BooleanLit(BooleanLit),
}

ast_node!(VarP, VarP);
ast_node!(ParenP, ParenP);
ast_node!(WildcardP, WildcardP);
ast_node!(AnnotationP, AnnotationP);

#[derive(PartialEq, Eq, Hash, Debug)]
#[repr(transparent)]
pub struct Pattern(SyntaxNode);

#[derive(Debug)]
pub enum PatternKind {
    VarP(VarP),
    ParenP(ParenP),
    WildcardP(WildcardP),
    AnnotationP(AnnotationP),
}

ast_node!(ParenTy, ParenTy);
ast_node!(IntTy, IntTy);
ast_node!(BoolTy, BoolTy);
ast_node!(FuncTy, FuncTy);

#[derive(PartialEq, Eq, Hash, Debug)]
#[repr(transparent)]
pub struct Type(SyntaxNode);

#[derive(Debug)]
pub enum TypeKind {
    ParenTy(ParenTy),
    IntTy(IntTy),
    BoolTy(BoolTy),
    FuncTy(FuncTy),
}

impl Type {
    fn cast(node: SyntaxNode) -> Option<Self> {
        if ParenTy::cast(node.clone()).is_some()
            || IntTy::cast(node.clone()).is_some()
            || BoolTy::cast(node.clone()).is_some()
            || FuncTy::cast(node.clone()).is_some()
        {
            Some(Type(node))
        } else {
            None
        }
    }

    pub fn kind(&self) -> TypeKind {
        ParenTy::cast(self.0.clone())
            .map(TypeKind::ParenTy)
            .or_else(|| IntTy::cast(self.0.clone()).map(TypeKind::IntTy))
            .or_else(|| BoolTy::cast(self.0.clone()).map(TypeKind::BoolTy))
            .or_else(|| FuncTy::cast(self.0.clone()).map(TypeKind::FuncTy))
            .unwrap()
    }

    pub fn range(&self) -> TextRange {
        self.0.text_range()
    }
}

impl Expr {
    fn cast(node: SyntaxNode) -> Option<Self> {
        if Var::cast(node.clone()).is_some()
            || Lambda::cast(node.clone()).is_some()
            || Application::cast(node.clone()).is_some()
            || IntLit::cast(node.clone()).is_some()
            || BooleanLit::cast(node.clone()).is_some()
        {
            Some(Expr(node))
        } else {
            None
        }
    }

    pub fn kind(&self) -> ExprKind {
        Var::cast(self.0.clone())
            .map(ExprKind::Var)
            .or_else(|| Lambda::cast(self.0.clone()).map(ExprKind::Lambda))
            .or_else(|| Application::cast(self.0.clone()).map(ExprKind::Application))
            .or_else(|| IntLit::cast(self.0.clone()).map(ExprKind::IntLit))
            .or_else(|| BooleanLit::cast(self.0.clone()).map(ExprKind::BooleanLit))
            .unwrap()
    }
}

impl Pattern {
    fn cast(node: SyntaxNode) -> Option<Self> {
        if VarP::cast(node.clone()).is_some()
            || ParenP::cast(node.clone()).is_some()
            || WildcardP::cast(node.clone()).is_some()
            || AnnotationP::cast(node.clone()).is_some()
        {
            Some(Pattern(node))
        } else {
            None
        }
    }

    pub fn kind(&self) -> PatternKind {
        VarP::cast(self.0.clone())
            .map(PatternKind::VarP)
            .or_else(|| ParenP::cast(self.0.clone()).map(PatternKind::ParenP))
            .or_else(|| WildcardP::cast(self.0.clone()).map(PatternKind::WildcardP))
            .or_else(|| AnnotationP::cast(self.0.clone()).map(PatternKind::AnnotationP))
            .unwrap()
    }
}
impl ParenTy {
    pub fn ty(&self) -> Type {
        self.0.children().find_map(Type::cast).unwrap()
    }
}

impl FuncTy {
    pub fn arg(&self) -> Type {
        self.0
            .children()
            .find_map(|x| {
                if x.kind() == SyntaxKind::TyArg {
                    x.children().find_map(Type::cast)
                } else {
                    None
                }
            })
            .unwrap()
    }

    pub fn res(&self) -> Type {
        self.0
            .children()
            .find_map(|x| {
                if x.kind() == SyntaxKind::TyRes {
                    x.children().find_map(Type::cast)
                } else {
                    None
                }
            })
            .unwrap()
    }
}

impl ParenP {
    pub fn pattern(&self) -> Pattern {
        self.0.children().find_map(Pattern::cast).unwrap()
    }
}
impl AnnotationP {
    pub fn pattern(&self) -> Pattern {
        self.0.children().find_map(Pattern::cast).unwrap()
    }

    pub fn ty(&self) -> Type {
        self.0.children().find_map(Type::cast).unwrap()
    }
}

impl IntLit {
    pub fn int(&self) -> i32 {
        self.0.text().to_string().parse().unwrap()
    }
}

impl BooleanLit {
    pub fn bool(&self) -> bool {
        self.0.text().to_string().parse().unwrap()
    }
}

impl Root {
    pub fn expr(&self) -> Option<Expr> {
        self.0.children().find_map(Expr::cast)
    }
}

pub fn first_token(node: &SyntaxNode) -> Option<SyntaxToken> {
    node.children_with_tokens().find_map(|node| {
        if node.kind() != SyntaxKind::Whitespace {
            node.as_token().cloned()
        } else {
            None
        }
    })
}

pub fn first_word(node: &SyntaxNode) -> Option<String> {
    node.children_with_tokens().find_map(|node| {
        if node.kind() == SyntaxKind::Ident {
            Some(node.as_token().unwrap().text().to_string())
        } else {
            None
        }
    })
}

impl Lambda {
    pub fn binder(&self) -> Option<AnnotationP> {
        self.0.children().find_map(AnnotationP::cast)
    }

    pub fn body(&self) -> Option<Expr> {
        self.0.children().find_map(Expr::cast)
    }
}

impl Var {
    pub fn name(&self) -> String {
        first_word(&self.0).unwrap()
    }
}

impl IntLit {
    pub fn value(&self) -> i32 {
        first_token(&self.0).unwrap().text().parse().unwrap()
    }
}

impl BooleanLit {
    pub fn value(&self) -> bool {
        first_token(&self.0).unwrap().kind() == SyntaxKind::TrueKw
    }
}

impl Application {
    pub fn func(&self) -> Expr {
        self.0.children().filter_map(Expr::cast).next().unwrap()
    }

    pub fn arg(&self) -> Expr {
        self.0.children().filter_map(Expr::cast).nth(1).unwrap()
    }
}

#[cfg(test)]
mod tests {
    use insta::{assert_debug_snapshot, glob};

    use super::*;

    #[test]
    fn parse_expr_test() {
        use std::fs;

        glob!("examples/expr/*.l3", |path| {
            let input = fs::read_to_string(path).unwrap();
            let parse = parse(&input);
            let node = parse.syntax();

            assert_debug_snapshot!((input, parse.errors, node));
        });
    }

    #[test]
    fn parse_type_test() {
        use std::fs;

        glob!("examples/types/*.l3", |path| {
            let input = fs::read_to_string(path).unwrap();
            let parse = parse_type(&input);
            let node = parse.syntax();

            assert_debug_snapshot!((input, parse.errors, node));
        });
    }

    #[test]
    fn parse_patterns_test() {
        use std::fs;

        glob!("examples/patterns/*.l3", |path| {
            let input = fs::read_to_string(path).unwrap();
            let parse = parse_pattern(&input);
            let node = parse.syntax();

            assert_debug_snapshot!((input, parse.errors, node));
        });
    }
}
