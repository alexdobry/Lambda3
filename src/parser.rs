use rowan::{Checkpoint, GreenNode, GreenNodeBuilder, Language};

use crate::cst;
use crate::lexer::SyntaxKind::{Eof, Error, Root};
use crate::lexer::{lex_str, SyntaxKind, Token};
use crate::syntax::{Lang, SyntaxNode};

mod expressions;
mod patterns;
mod types;

/// The parse results are stored as a "green tree".
/// We'll discuss working with the results later
#[derive(Debug)]
pub struct Parse {
    pub green_node: GreenNode,
    #[allow(unused)]
    pub errors: Vec<String>,
}

pub struct Parser {
    /// input tokens, including whitespace,
    /// in *reverse* order.
    tokens: Vec<(Vec<Token>, Token)>,
    /// the in-progress tree.
    builder: GreenNodeBuilder<'static>,
    /// the list of syntax errors we've accumulated
    /// so far.
    errors: Vec<String>,
}

impl Parse {
    pub fn syntax(&self) -> SyntaxNode {
        SyntaxNode::new_root(self.green_node.clone())
    }

    pub fn root(&self) -> cst::Root {
        cst::Root::cast(self.syntax()).unwrap()
    }

    pub fn debug_tree(&self) -> String {
        let syntax_node = SyntaxNode::new_root(self.green_node.clone());
        let formatted = format!("{:#?}", syntax_node);

        // We cut off the last byte because formatting the SyntaxNode adds on a newline at the end.
        formatted[0..formatted.len() - 1].to_string()
    }
}

impl Parser {
    fn new(tokens: Vec<(Vec<Token>, Token)>) -> Self {
        Parser {
            tokens,
            builder: GreenNodeBuilder::new(),
            errors: Vec::new(),
        }
    }

    fn start_node(&mut self, kind: SyntaxKind) {
        self.builder.start_node(Lang::kind_to_raw(kind));
    }

    fn start_node_at(&mut self, checkpoint: Checkpoint, kind: SyntaxKind) {
        self.builder
            .start_node_at(checkpoint, Lang::kind_to_raw(kind));
    }

    fn finish_node(&mut self) {
        self.builder.finish_node();
    }

    fn finish_at(&mut self, c: Checkpoint, kind: SyntaxKind) {
        self.start_node_at(c, kind);
        self.finish_node();
    }

    fn checkpoint(&self) -> Checkpoint {
        self.builder.checkpoint()
    }

    fn report_error(&mut self, msg: String) {
        self.start_node(Error);
        self.errors.push(msg);
        self.finish_node();
    }

    fn nth(&self, n: usize) -> SyntaxKind {
        let len = self.tokens.len();
        if n >= len {
            SyntaxKind::Eof
        } else {
            self.tokens[len - n - 1].1 .0
        }
    }

    /// Peek at the first unprocessed token
    fn current(&self) -> SyntaxKind {
        self.nth(0)
    }

    fn at(&self, kind: SyntaxKind) -> bool {
        self.current() == kind
    }

    fn eat(&mut self, kind: SyntaxKind) -> bool {
        if !self.at(kind) {
            return false;
        }
        self.bump_any();
        true
    }

    /// Advance one token, adding it to the current branch of the tree builder.
    fn bump_any(&mut self) {
        if self.at(Eof) {
            return;
        }
        let (leading, (kind, text)) = self.tokens.pop().unwrap();

        for (kind, text) in leading {
            self.builder.token(Lang::kind_to_raw(kind), text.as_str());
        }

        self.builder.token(Lang::kind_to_raw(kind), text.as_str());
    }

    fn bump(&mut self, kind: SyntaxKind) {
        assert!(self.eat(kind))
    }
}

pub fn parse(text: &str) -> Parse {
    let mut tokens = lex_str(text);
    tokens.reverse();
    let mut p = Parser::new(tokens);

    p.start_node(Root);
    let _ = expressions::expr(&mut p);
    if !p.at(SyntaxKind::Eof) {
        p.start_node(Error);
        while !p.at(Eof) {
            p.bump_any()
        }
        p.finish_node();
        p.errors.push(format!("Unexpected token {:?}", p.current()))
    }
    p.finish_node();

    Parse {
        green_node: p.builder.finish(),
        errors: p.errors,
    }
}

pub fn parse_type(text: &str) -> Parse {
    let mut tokens = lex_str(text);
    tokens.reverse();
    let mut p = Parser::new(tokens);

    p.start_node(Root);
    let _ = types::typ(&mut p);
    if !p.at(SyntaxKind::Eof) {
        p.start_node(Error);
        while !p.at(Eof) {
            p.bump_any()
        }
        p.finish_node();
        p.errors.push(format!("Unexpected token {:?}", p.current()))
    }
    p.finish_node();

    Parse {
        green_node: p.builder.finish(),
        errors: p.errors,
    }
}

pub fn parse_pattern(text: &str) -> Parse {
    let mut tokens = lex_str(text);
    tokens.reverse();
    let mut p = Parser::new(tokens);

    p.start_node(Root);
    let _ = patterns::parse_pattern(&mut p);
    if !p.at(SyntaxKind::Eof) {
        p.start_node(Error);
        while !p.at(Eof) {
            p.bump_any()
        }
        p.finish_node();
        p.errors.push(format!("Unexpected token {:?}", p.current()))
    }
    p.finish_node();

    Parse {
        green_node: p.builder.finish(),
        errors: p.errors,
    }
}
