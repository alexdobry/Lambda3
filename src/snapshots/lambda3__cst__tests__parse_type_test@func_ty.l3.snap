---
source: src/cst.rs
expression: "(input, parse.errors, node)"
input_file: src/examples/types/func_ty.l3

---
(
    "Int->Bool->Int",
    [],
    Root@0..14
      FuncTy@0..14
        TyArg@0..3
          IntTy@0..3
            Int@0..3 "Int"
        Arrow@3..5 "->"
        TyRes@5..14
          FuncTy@5..14
            TyArg@5..9
              BoolTy@5..9
                Bool@5..9 "Bool"
            Arrow@9..11 "->"
            TyRes@11..14
              IntTy@11..14
                Int@11..14 "Int"
    ,
)
